src/key_store/mod.rs

```diff
@@ -0,0 +1,223 @@
+/// `KeyPackage` instances.
+pub struct KeyStore {
+    // Map from signature public keys to credential bundles
+    credential_bundles: RwLock<HashMap<SignaturePublicKey, CredentialBundle>>,
+    init_key_package_bundles: RwLock<HashMap<Vec<u8>, KeyPackageBundle>>,
+}
+
+/// This guard struct for a `CredentialBundle` implements `Deref`, such that the
+/// underlying `RwLock<CredentialBundle>` can be obtained for read or write
+/// access to the credential.
+pub struct CbGuard<'a> {
+    cbs: RwLockReadGuard<'a, HashMap<SignaturePublicKey, CredentialBundle>>,
+    index: &'a SignaturePublicKey,
+}
+
+use std::ops::Deref;
+
+impl<'b> Deref for CbGuard<'b> {
+    type Target = CredentialBundle;
+
+    fn deref(&self) -> &CredentialBundle {
+        // We can unwrap here, as we checked if the entry is present before
+        // creating the guard. Also, since we hold a read lock on the `HashMap`,
+        // the entry can't have been removed in the meantime.
+        self.cbs.get(self.index).unwrap()
+    }
+}
+
+impl KeyStore {
+    /// Retrieve a `KeyPackageBundle` from the key store given the hash of the
+    /// corresponding `KeyPackage`. This removes the `KeyPackageBundle` from the
+    /// store. Returns an error if no `KeyPackageBundle` can be found
+    /// corresponding to the given `KeyPackage` hash.
+    pub fn take_key_package_bundle(&self, kp_hash: &[u8]) -> Option<KeyPackageBundle> {
+        // We unwrap here, because the two functions claiming a write lock on
+        // `init_key_package_bundles` (this one and `generate_key_package`) only
+        // hold the lock very briefly and should not panic during that period.
+        let mut kpbs = self.init_key_package_bundles.write().unwrap();
+        kpbs.remove(kp_hash)
+    }
+
+    /// Retrieve a `CbGuard` from the key store given the `SignaturePublicKey`
+    /// of the corresponding `Credential`. The `CbGuard` can be dereferenced to
+    /// obtain an `RwLock` on the desired `CredentialBundle`. Returns an error
+    /// if no `CredentialBundle` can be found corresponding to the given
+    /// `SignaturePublicKey`.
+    pub(crate) fn get_credential_bundle<'key_store>(
+        &'key_store self,
+        signature_public_key: &'key_store SignaturePublicKey,
+    ) -> Option<CbGuard> {
+        let cbs = self.credential_bundles.read().unwrap();
+        if !cbs.contains_key(signature_public_key) {
+            return None;
+        }
+        Some(CbGuard {
+            cbs,
+            index: signature_public_key,
+        })
+    }
+
+    /// Generate a fresh `KeyPackageBundle` with the given parameters, store it
+    /// in the `KeyStore` and return the corresponding `KeyPackage`. Throws an
+    /// error if no `CredentialBundle` can be found in the `KeyStore`
+    /// corresponding to the given `Credential` or if an error occurred during
+    /// the creation of the `KeyPackageBundle`.
+    pub fn generate_key_package(
```

Comment:
This generates a `KeyPackageBundle` not a `key_package`.

Reply:
I think this is fine though, because the bundle is now only stored within the KeyStore. From an outside perspective, we only ever deal with key packages. The return type is consistent with the function name and the caller doesn't need to know what's happening internally.

Reply:
It's in line with the other functions. But it's still not doing what it says, i.e. it's not generating only a key package.